File: machine/class/logger.cpp
#include "memory.hpp"

#include "class/class.hpp"
#include "class/logger.hpp"
#include "class/string.hpp"

#include "logger.hpp"

namespace rubinius {
  using namespace utilities;

  void Logger::bootstrap(STATE) {
    GO(logger).set(state->memory()->new_class<Class, Logger>(
          state, G(rubinius), "Logger"));
  }

  Logger* Logger::create(STATE) {
    return Logger::allocate(state, G(logger));
  }

  Logger* Logger::allocate(STATE, Object* self) {
    return state->memory()->new_object<Logger>(state, as<Class>(self));
  }

  Object* Logger::write(STATE, String* message) {
    logger::write(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::fatal(STATE, String* message) {
    logger::fatal(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::error(STATE, String* message) {
    logger::error(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::warn(STATE, String* message) {
    logger::warn(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::info(STATE, String* message) {
    logger::info(format()->c_str(state), message->c_str(state));
    return cNil;
  }

  Object* Logger::debug(STATE, String* message) {
    logger::debug(format()->c_str(state), message->c_str(state));
    return cNil;
  }
}

File: core/logger.rb
module Rubinius
  class Logger
    attr_accessor :name
    attr_accessor :format

    def self.allocate
      Rubinius.primitive :logger_allocate
      raise PrimitiveFailure, "Rubinius::Logger.allocate failed"
    end

    def self.system
      @system_logger ||= new "system"
    end

    def self.log_exception(message, e)
      log = system

      exception = e
      exceptions = []

      while exception
        exceptions.unshift exception
        exception = exception.cause
      end

      color = STDERR.tty?

      while exception = exceptions.shift
        if custom_trace = exception.custom_backtrace
          msg = "User-defined backtrace:"
          log.error msg
          STDERR.puts msg

          custom_trace.reverse_each do |line|
            log.error line
            STDERR.puts line
          end

          msg = "Backtrace: "
          log.error msg
          STDERR.puts "\n", msg, "\n"
        end

        exception.backtrace.reverse_each { |line| log.error line }
        STDERR.puts exception.awesome_backtrace.show("\n", color)

        msg = "#{exception.message} (#{exception.class})"
        log.error msg
        STDERR.puts "\n", msg

        unless exceptions.empty?
          msg = "Causing:"
          log.error msg
          STDERR.puts "\n", msg, "\n"
        end
      end

      log.error message
      STDERR.puts "\n", message
    end

    def initialize(name)
      @name = name
      @format = "#{name}: %s"
    end

    private :initialize

    def write(message)
      Rubinius.primitive :logger_write
      raise PrimitiveFailure, "Rubinius::Logger#write primitive failed"
    end

    def fatal(message)
      Rubinius.primitive :logger_fatal
      raise PrimitiveFailure, "Rubinius::Logger#fatal primitive failed"
    end

    def error(message)
      Rubinius.primitive :logger_error
      raise PrimitiveFailure, "Rubinius::Logger#error primitive failed"
    end

    def warn(message)
      Rubinius.primitive :logger_warn
      raise PrimitiveFailure, "Rubinius::Logger#warn primitive failed"
    end

    def info(message)
      Rubinius.primitive :logger_info
      raise PrimitiveFailure, "Rubinius::Logger#info primitive failed"
    end

    def debug(message)
      Rubinius.primitive :logger_debug
      raise PrimitiveFailure, "Rubinius::Logger#debug primitive failed"
    end
  end
end
